tm2/pkg/std/memfile.go

```diff
@@ -2,99 +2,231 @@ package std
+		lname := strings.ToLower(mfile.Name)
+		if _, exists := uniq[lname]; exists {
+			return fmt.Errorf("duplicate file name %q", lname)
+		}
+		uniq[lname] = struct{}{}
+	}
+	return nil
 }
 
-const pathLengthLimit = 256
-
-var (
-	rePkgName      = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)
-	rePkgOrRlmPath = regexp.MustCompile(`^([a-zA-Z0-9-]+\.)*[a-zA-Z0-9-]+\.[a-zA-Z]{2,}\/(?:p|r)(?:\/_?[a-z]+[a-z0-9_]*)+$`)
-	reFileName     = regexp.MustCompile(`^([a-zA-Z0-9_]*\.[a-z0-9_\.]*|LICENSE|README)$`)
-)
+// Sort files; a MemPackage with unordered files is in valid.
+func (mpkg *MemPackage) Sort() {
+	slices.SortFunc(mpkg.Files, func(a, b *MemFile) int {
+		return strings.Compare(a.Name, b.Name)
+	})
+}
 
-// path must not contain any dots after the first domain component.
-// file names must contain dots.
-// NOTE: this is to prevent conflicts with nested paths.
-func (mempkg *MemPackage) Validate() error {
-	// add assertion that MemPkg contains at least 1 file
-	if len(mempkg.Files) <= 0 {
-		return fmt.Errorf("no files found within package %q", mempkg.Name)
+// Return the named file or none if it doesn't exist.
+func (mpkg *MemPackage) GetFile(name string) *MemFile {
+	for _, mfile := range mpkg.Files {
+		if mfile.Name == name {
+			return mfile
+		}
 	}
+	return nil
+}
 
-	if len(mempkg.Path) > pathLengthLimit {
-		return fmt.Errorf("path length %d exceeds limit %d", len(mempkg.Path), pathLengthLimit)
-	}
+func (mpkg *MemPackage) appendFile(mfile *MemFile) {
+	mpkg.Files = append(mpkg.Files, mfile)
+}
 
-	if !rePkgName.MatchString(mempkg.Name) {
-		return fmt.Errorf("invalid package name %q, failed to match %q", mempkg.Name, rePkgName)
+// Creates a new MemFile and appends without validation.
+func (mpkg *MemPackage) NewFile(name string, body string) (mfile *MemFile) {
+	mfile = &MemFile{
+		Name: name,
+		Body: body,
 	}
+	mpkg.appendFile(mfile)
+	return
+}
 
-	if !rePkgOrRlmPath.MatchString(mempkg.Path) {
-		return fmt.Errorf("invalid package/realm path %q, failed to match %q", mempkg.Path, rePkgOrRlmPath)
-	}
-	// enforce sorting files based on Go conventions for predictability
-	sorted := sort.SliceIsSorted(
-		mempkg.Files,
-		func(i, j int) bool {
-			return mempkg.Files[i].Name < mempkg.Files[j].Name
-		},
-	)
-	if !sorted {
-		return fmt.Errorf("mempackage %q has unsorted files", mempkg.Path)
+// Writes to existing file or creates a new one.
+func (mpkg *MemPackage) SetFile(name string, body string) *MemFile {
+	for _, mfile := range mpkg.Files {
+		if mfile.Name == name {
+			mfile.Body = body
+			return mfile
+		}
 	}
+	return mpkg.NewFile(name, body)
+}
 
-	var prev string
-	for i, file := range mempkg.Files {
-		if !reFileName.MatchString(file.Name) {
-			return fmt.Errorf("invalid file name %q, failed to match %q", file.Name, reFileName)
-		}
-		if i > 0 && prev == file.Name {
-			return fmt.Errorf("duplicate file name %q", file.Name)
+// Returns true if zero.
+// XXX Reconsider renaming all IsEmpty() to IsZero(), each carefully.
```

Comment:
This is definitely an IsZero, while IsEmpty could simply be `len(mpkg.Files) == 0` without checking the package name.
I'll start using IsZero instead of IsEmpty wherever it makes sense.